File: src/C/cache.c
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0.  If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * (c) 2020- MonetDB Solutions B.V.
 *
 * An embedded application can act a cache with a remote server.
 * For this we need two database objects, one :inmemory: and a link to the remote.
 *
 * For an explanation of the command arguments see ...
 */

#include "monetdbe.h"
#include <stdlib.h>
#include <stdio.h>
#include <inttypes.h>

#define error(msg) {fprintf(stderr, "Failure: %s\n", msg); exit -1;}


int
main(void)
{
    monetdbe_database local = NULL, remote = NULL;
    monetdbe_result *data, *schema, *result;
    monetdbe_column *schemas;

    monetdbe_open(&remote, "monetdb://localhost:5000/sf1?user=monetdb&password=monetdb", NULL);  
    monetdbe_query(remote, "call sys.describe(\"sys\",\"lineitem\"", &schema); 
    monetdbe_query(remote, "select line_no from sf1 where line_no < 10", &data);
    
    if (remote->error || schema->error || data->error) 
        error("Could not access the remote database");

    // query returns a single string
    monetdb_result_fetch(schema, &schemas, 0);
    schemadef = (char*) schemas->data[0];

    monetdb_cleanup_result(remote, schema);

    // store the result in an :inmemory: structure
    monetdbe_open(&local, NULL, NULL);  
    monetdbe_query(local, schemadef, &schema);
    monetdb_result_fetch(schema, &schemas, 0);

    /* need to fetch the result first */
    monetdbe_append(local, "sys","lineitem", data, &result);

    if( local->error || schema->error || result->error )
        error("Construction of the local cache failed\n");

	printf("Obtained %d tuples from the remote", result->affectedrows);
}

File: python/cache.py
"""
 This Source Code Form is subject to the terms of the Mozilla Public
 License, v. 2.0.  If a copy of the MPL was not distributed with this
 file, You can obtain one at http://mozilla.org/MPL/2.0/.

 (c) 2020- MonetDB Solutions B.V.

 An embedded application can act a cache with a remote server.
 For this we need two database objects, one :memory: and an URL link to the remote.

 For an explanation of the command arguments see MonetDBe/Python documentation
"""
import monetdbe

if __name__ == "__main__":
    print('This experiment is on hold until we support remote connections')
    exit(-1)
    
    try:
        remote = monetdbe.connect("monetdb://localhost:5000/sf1?user=monetdb&password=monetdb")
        curr = remote.cursor()
        schema = curr.execute("call sys.describe(\"sys\",\"lineitem\"")
        data = curr.execute("select line_no from sf1 where line_no < 10")
        
        if remote.error or schema.error or data.error:
            print("Could not access the remote database")
            exit (-1)

        schemadef = schema.fetchone()

        # store the result in an :memory: structure
        local = monetdbe.connect(None)
        curl = local.cursor()
        schema = curl.execute(schemadef)
        curl.append("sys","lineitem", data)

        if local.error or schema.error or result.error:
            print("Construction of the local cache failed\n")
            exit (-1)

        print("Obtained tuples from the remote")

    except monetdbe.exceptions.OperationalError as msg:
        print(f"Error encountered {msg}")
